### PR TITLE
Added support for slicing scenarios

### PR DESCRIPTION
**Pull request by snorfalorpagus**
Added support for slicing scenarios, in order to run only a subset.

Example usage:

```
scenario_weather = Scenario(model, size=200)
scenario_demand = Scenario(model, size=3)
scenario_weather.slice = slice(0, 50, 1) # only the first 50 weather scenarios
scenario_demand.slice = slice(2, 3, 1) # only the last demand scenario
```

This partially addresses #385. Not implemented is JSON support (is this needed?) or support for arbitrary combinations.

I'd like to use this next week. We can add the other bits as needed.

### CODE REVIEW

File: pywr/_core.pyx [hunk 1]
```diff
@@ -64,19 +46,22 @@ cdef class ScenarioCollection:
...
-            combinations = [ScenarioIndex(0, np.array([0], dtype=np.int32)), ]
+            # model has no scenarios defined, implicitly has 1 scenario of size 1
+            combinations = [ScenarioIndex(0, np.array([0], dtype=np.int32))]
         else:
-            combinations = []
-            for i, indices in enumerate(product([sc._size for sc in self._scenarios])):
-                combinations.append(ScenarioIndex(i, indices))
-        self.combinations = combinations
+            # product of all scenarios, taking into account Scenario.slice
+            iter = itertools.product(*[range(scenario._size)[scenario.slice] if scenario.slice else range(scenario._size) for scenario in self._scenarios])
```
Comment (jetuk) [inline pywr/_core.pyx]: This line is nuts; I like it.

**Comment by jetuk** [comment]
I think this is fine. I might do the explicit list on top this weekend.

JSON support would be good. I guess it'll be forgotten about if not done now.

**Comment by jetuk** [comment]
Did the user defined combinations, but no JSON support.

**Comment by snorfalorpagus** [comment]
I'll review this tomorrow.

**Pull request merged**